libmamba/src/core/compression.hpp

```diff
@@ -14,6 +14,42 @@
 
 namespace mamba
 {
+
+    class CompressionStream
+    {
+    public:
+
+        using writer = std::function<size_t(char*, size_t)>;
+
+        virtual ~CompressionStream() = default;
+
+        CompressionStream(const CompressionStream&) = delete;
+        CompressionStream& operator=(const CompressionStream&) = delete;
+        CompressionStream(CompressionStream&&) = delete;
+        CompressionStream& operator=(CompressionStream&&) = delete;
+        
+        size_t write(char* in, size_t size);
+
+    protected:
+
+        CompressionStream(writer&& func);
+
+        size_t invoke_writer(char* in, size_t size);
+
+    private:
+
+        virtual size_t write_impl(char* in, size_t size) = 0;
+
+        writer m_writer;
+    };
+
+    std::unique_ptr<CompressionStream> make_compression_stream(
+        const std::string& url,
+        CompressionStream::writer&& func
+    );
+
+    // TODO: remove the following when switching to new CompressionStream
```

Comment:
remove ZstdStream?

Reply:
Yes it will be replaced with the hidden ZstdCompressionStream defined in the cpp.